pipeline/trusted-resources/pkg/trustedtask/trustedpipelinerun.go

```diff
@@ -0,0 +1,188 @@
+		return errs
+	}
+	return nil
+}
+
+// SetDefaults is not used.
+func (pr *TrustedPipelineRun) SetDefaults(ctx context.Context) {
+}
+
+func (pr *TrustedPipelineRun) verifyPipelineRun(
+	ctx context.Context,
+	k8sclient kubernetes.Interface,
+	tektonClient versioned.Interface,
+) (errs *apis.FieldError) {
+	logger := logging.FromContext(ctx)
+	logger.Info("Verifying PipelineRun")
+
+	serviceAccountName := os.Getenv("WEBHOOK_SERVICEACCOUNT_NAME")
+	if serviceAccountName == "" {
+		serviceAccountName = "tekton-verify-task-webhook"
+	}
+
+	// for inline pipeline we verify the tasks
+	if pr.Spec.PipelineRef == nil {
+		for _, t := range pr.Spec.PipelineSpec.Tasks {
+			if t.TaskRef == nil {
+				continue
+			}
+
+			fn, err := taskrun.GetTaskFunc(ctx, k8sclient, tektonClient, t.TaskRef, pr.Namespace, serviceAccountName)
+			if err != nil {
+				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+			}
+
+			if err := verifyTask(ctx, t.TaskRef.Name, k8sclient, fn); err != nil {
+				return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+			}
+		}
+		return nil
+	}
+
+	cfg := config.FromContextOrDefaults(ctx)
+	cfg.FeatureFlags.EnableTektonOCIBundles = true
+	ctx = config.ToContext(ctx, cfg)
+
+	fn, err := resources.GetPipelineFunc(ctx, k8sclient, tektonClient, &pr.PipelineRun)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+	resolvedPipeline, err := fn(ctx, pr.Spec.PipelineRef.Name)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+
+	pipeline, signature, err := preparePipeline(resolvedPipeline)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "spec", "PipelineRef")
+	}
+
+	v, err := verifier(ctx, pipeline.ObjectMeta.Annotations, k8sclient)
+	if err != nil {
+		return apis.ErrGeneric(err.Error(), "PipelineRef")
+	}
+
+	if err := VerifyInterface(ctx, pipeline, v, signature); err != nil {
+		return apis.ErrGeneric(err.Error(), "PipelineRef")
+	}
+
+	for _, t := range pipeline.Spec.Tasks {
+		if t.TaskRef == nil {
+			continue
+		}
+
+		fn, err := taskrun.GetTaskFunc(ctx, k8sclient, tektonClient, t.TaskRef, pr.Namespace, serviceAccountName)
+		if err != nil {
+			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+		}
+
+		if err := verifyTask(ctx, t.TaskRef.Name, k8sclient, fn); err != nil {
+			return apis.ErrGeneric(err.Error(), "spec", "taskRef")
+		}
+
+	}
+
+	return nil
+}
+
+// preparePipeline will convert the taskobject to task and extract the signature.
+func preparePipeline(p v1beta1.PipelineObject) (v1beta1.Pipeline, []byte, error) {
+	pipeline := v1beta1.Pipeline{}
+	pipeline.TypeMeta = metav1.TypeMeta{
+		APIVersion: "tekton.dev/v1beta1",
+		Kind:       "Pipeline"}
+	pipeline.Spec = p.PipelineSpec()
+
+	// exclude the fields populated by system.
+	pipeline.Name = p.PipelineMetadata().Name
+	pipeline.GenerateName = p.PipelineMetadata().GenerateName
+	pipeline.Namespace = p.PipelineMetadata().Namespace
+
+	pipeline.Labels = make(map[string]string)
```

Comment:
Should we do something similar to https://github.com/tektoncd/experimental/blob/de213d0a1aafb75c3baf0266e886534d2dc0b2e5/pipeline/trusted-resources/pkg/trustedtask/trustedtaskrun.go#L142-L147 here?